### PR TITLE
Add deprecate attribute module and spec

### PR DESCRIPTION
**Pull request by yieldjessyield**
/domain @pat-whitrock 
/platform @smudge 

This is my first pass at pulling our `DeprecatedFirstName` module from our app into a more generalized `DeprecateAttribute` module in Uncruft. I've heard from a few engineers that the logic in `DeprecatedFirstName` would be useful in other deprecation use cases. Curious what folks think, if we are able to merge this change then my next step would be to update our Retail User class to use Uncruft's `DeprecateAttribute` and remove our custom built `DeprecatedFirstName` module.

All specs are passing locally.

Paired with @pat-whitrock on this!

### CODE REVIEW

**Comment by pat-whitrock** [comment]
<< domain

@smudge I'm curious if you think this belongs more in Uncruft or Runtime. Uncruft felt like a reasonable home because deprecation warnings raising in development and test is fundamental to `DeprecateAttribute`'s utility. But shorthand for deprecating a method isn't necessarily tied to Uncruft.

This feels like it would be expanding the scope of Uncruft's current purpose:

> A library to assist with clearing out Rails deprecation warnings and upgrading Rails versions.

`DeprecateAttribute` broadens that scope a bit to something like this:

> A library to assist with clearing out deprecation warnings and upgrading frameworks.

`DeprecateAttribute` allows us to more formally treat our own applications' APIs as library code and other Betterment engineers as consumers of those libraries. That may feel like an odd way to think of our own systems and coworkers, but our full stack teams are both specialized in their respective domains and empowered to work across those boundaries in a way that makes it difficult to migrate an API one step forward without adjacent teams pulling it two steps back. This feels particularly difficult in a monorepo, as opposed to a system with a more rigid structure (like SOA) that may inherently communicate the preferred API moving forward.

I'm curious if this feels like something that would help you, as someone who presumably spends a disproportionate amount of time introducing broad changes across our systems that are being constantly altered by engineers who have several other priorities on their minds. I could see relying more on deprecation warnings and Uncruft within our internal APIs providing the same two main benefits we've enjoyed for Rails upgrades:
- Introduce friction when attempting to introduce a new dependency on a deprecated API
- Provide a to-do list of what needs to be migrated to the new system

Together, that should reduce the total amount of work needed to be done and make it easier to delegate that work across the org.

My main concern with this approach is if filling the `deprecation.ignore` with both external (Rails version upgrades) and internal (e.g. legal name migration) deprecation warnings would make it difficult to differentiate between the work necessary to complete any individual initiative. Maybe that wouldn't actually be too difficult to parse out, but if anyone shares that concern we could consider tracking deprecation warnings in separate files for each initiative (e.g. `legal_name_deprecations.ignore`).

File: lib/uncruft/deprecate_attribute.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+module Uncruft
+  module DeprecateAttribute
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def deprecate_attribute(attribute, aliased_attribute: nil, message:)
```
Comment (pat-whitrock) [inline lib/uncruft/deprecate_attribute.rb:6]: This could be considered a v2 feature since I believe the majority of potential use cases we've seen have been AR attributes, but I could see a `deprecate_method` that doesn't assume there's both a reader and writer being useful.

File: spec/uncruft/deprecate_attribute_spec.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
+require 'spec_helper'
+
+RSpec.describe Uncruft::DeprecateAttribute do
+  let(:my_name) { "Jess" }
+
+  subject { klass.new }
+
+  describe '#deprecate_attribute' do
```
Comment (pat-whitrock) [inline spec/uncruft/deprecate_attribute_spec.rb:8]: Nit: `deprecate_attribute` is a class method, so `.deprecate_attribute` would be a more idiomatic RSpec describe

File: lib/uncruft/deprecate_attribute.rb (hunk 1 above)
Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: Agreed. Not necessary as part of this initial release, but I could see `deprecate_attribute` being a convenience wrapper for the combination `deprecate_reader` and `deprecate_writer`.

File: spec/uncruft/deprecate_attribute_spec.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        expect(subject.first_name = my_name).to eq my_name
+
+        expect(ActiveSupport::Deprecation).to receive(:warn)
+          .with("Please stop using this attribute, use this other attribute instead!")
+
+        expect(subject.first_name).to eq my_name
+      end
+
+      it 'returns attribute when setting and getting aliased_attribute' do
```
Comment (pat-whitrock) [inline spec/uncruft/deprecate_attribute_spec.rb:34]: Can we also test that the deprecation warnings _don't_ happen in this spec?

File: spec/uncruft/deprecate_attribute_spec.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+          attr_accessor :first_name
+
+          deprecate_attribute(:first_name,
+                              aliased_attribute: :legal_first_name,
+                              message: "Please stop using this attribute, use this other attribute instead!")
+        end
+      end
+
+      it 'applies deprecation warning when setting and getting deprecated attribute' do
+        expect(ActiveSupport::Deprecation).to receive(:warn)
```
Comment (pat-whitrock) [inline spec/uncruft/deprecate_attribute_spec.rb:23]: If we're going to test the writer and reader together in a single spec, we probably want to be [more specific](https://relishapp.com/rspec/rspec-mocks/docs/setting-constraints/receive-counts)

```ruby
expect(...).to receive(...).exactly(n).times
```

Or we could break this up into two separate specs and set the reader spec up with `instance_variable_set`.

File: spec/uncruft/deprecate_attribute_spec.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+          attr_accessor :first_name
+
+          deprecate_attribute(:first_name,
+                              message: "Please stop using this attribute!")
+        end
+      end
+
+      it 'applies deprecation warning when setting and getting deprecated attribute' do
+        expect(ActiveSupport::Deprecation).to receive(:warn)
```
Comment (pat-whitrock) [inline spec/uncruft/deprecate_attribute_spec.rb:54]: Same note as the spec above WRT receive counts or breaking this into two specs

File: lib/uncruft/deprecate_attribute.rb (hunk 1 above)
Comment (pat-whitrock) [inline lib/uncruft/deprecate_attribute.rb:6]: Would be interested to hear feedback on `aliased_attribute` naming from someone with fresh 👀

Comment (pat-whitrock) [reply lib/uncruft/deprecate_attribute.rb:6]: I'm skeptical `reader` and `writer` could exclusively cover all cases for this. I could see us deprecating methods that don't fit into that framing (e.g. a balance method that takes a date as an argument).

**Comment by pat-whitrock** [comment]
domain TAFN

I don't wanna get ahead of ourselves, but I'd be also interested in shipping this with a shared example. Could also theoretically be a v2 feature.

File: lib/uncruft/deprecate_attribute.rb (hunk 1 above)
Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: Fair enough! There's probably a generic "this method is deprecated" method (regardless of # of args or whether it is a reader/writer) that is the superset of all of the above.

Comment (smudge) [inline lib/uncruft/deprecate_attribute.rb:6]: I think it would be really cool if we made `message` optional and auto-generated a method based on the inputs.

Basically, if you don't specify an alias, we output a simple method on read or on write:
- `[attribute] is deprecated. Please avoid using it.`
- `[attribute]= is deprecated. Please avoid using it.`

And if you do specify an alias, we can say:
- `[attribute] is deprecated. Please use [aliased_attribute] instead.`
- `[attribute]= is deprecated. Please use [aliased_attribute]= instead.`

Thoughts?

Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: I'm wondering whether this aliasing feature will cause confusion, and if there's anything we can do to anticipate that.

Basically, I think there are two ways of reading `aliased_attribute`:

1. "Please alias this for me now" (the current implementation)
2. "I have already implemented the other/aliased method but I want to indicate that this method is deprecated in favor of that other one."

In the case of number 2, it would result in someone potentially overwriting a method that they already defined, and unexpected behavior in general.

I also left [a note about potentially auto-generating the message](https://github.com/Betterment/uncruft/pull/4/files#r461027022), but I think that opens up a can of worms when it comes to the "use X insetad" case, because the ability to specify an alternate, non-deprecated method (without having this framework auto-alias it) feels pretty reasonable in that case.

**Comment by smudge** [comment]
<< platform TAFN - also don't want to explode the scope of this, mostly just wanting to prompt some conversations and think about the direction we _might_ want to go with this!

But overall I love it. I definitely think it fits within the goals of this project. I'd be totally happy to ship this as-is (and make breaking changes to it later if we see fit), but the one thing I do think we need still is an update to the README indicating that this new feature exists and outlining how to use it!

(While we're at it, we could also make the README less Rails-centric and change the description of the project to look more like what @pat-whitrock proposed above. But that can be a separate PR too, either way! 👍 )

File: lib/uncruft/deprecate_attribute.rb (hunk 1 above)
Comment (pat-whitrock) [reply lib/uncruft/deprecate_attribute.rb:6]: I'm concerned that the risk of someone not realizing they can provide a custom message when they should outweighs the benefit of saving yourself a minute or so of crafting something simple.

From the examples where I would have liked to use this, I don't think an auto-generated message would capture everything I want to convey.

- `Please avoid using it` is probably not enough context in a lot of situations. I would imagine in most cases where we're deprecating something, there's a preferred alternative or some sort of explanation for how to reframe what we're trying to accomplish.
- Sometimes we split a single concept into two or more separate concepts, so the alias may always be the correct alternative. For example, if we had a `name` attribute and wanted to split it into `first_name` and `nickname`, we don't want to say `Please  use first_name`. We want to say something like `Please use first_name in formal contexts and nickname in casual contexts.`

Do you share those concerns? If we want to provide some default option, maybe you could pass `message: :default`?

Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: Yeah `message: :default` could work. I mean I could go either way on this. I do agree that in a lot of cases it makes sense to specify a custom message. However, it's also possible to use `ActiveSupport::Deprecation.warn` directly without much additional effort beyond what this mixin provides. That sort of led me to think that we want to focus on more targeted use cases, like "attribute X has been renamed to Y", etc, and the more targeted the use case, the more natural it feels to auto-generate a message.

Basically, since we're intentionally constraining the problem, we're inherently starting to classify the ways in which a method might be deprecated, so in my mind I'd expect us to converge on similar messages that correspond to each of those classifications.

Comment (pat-whitrock) [reply lib/uncruft/deprecate_attribute.rb:6]: 👍 that's a very valid concern. The tense there is maybe part of the issue. Uncruft doesn't want you to provide the name of an alias you created - it wants to create the alias for you.

I also wonder if we should be less focused on what we're going to do internally with the argument you provide (create an alias of the deprecated method by this name) and more focused on how it will alter the interface of the class (add a "safe", deprecation-less version of the deprecated method by this name).

What about this?

```ruby
def deprecate_attribute(attribute, safe_name: nil, message:)
```

Comment (pat-whitrock) [reply lib/uncruft/deprecate_attribute.rb:6]: Yeah, fair enough. If we see a strong majority of cases that fall within the same patterns, I'll be a lot more convinced. With the goal of shipping this feature quickly and learning more about how folks decide to use it, maybe we should bookmark this as v2 idea to revisit?

Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: `safe_name` is even less clear to me though, because I still read it as "deprecate 'X', because the safe name (that I have already implemented) is 'Y'"

What if we turn the intent entirely on its head and make the API look like this:

```ruby
alias_attribute :nickname, :name, deprecation: "`name` is deprecated. Please use `nickname` instead"
```

Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: (e.g. extend the existing alias method with a new deprecation arg)

Comment (pat-whitrock) [reply lib/uncruft/deprecate_attribute.rb:6]: Oh, interesting. I like that. We'd be giving up a couple things, though:

- Deprecating non-attribute methods
- DSL for deprecating non-aliased attributes

I'm more concerned about the first one, as I can see aliased deprecation use cases for non-ActiveModel-attribute methods (e.g. updated balance calculation, Her model attributes, etc).

Less concerned about the second case, but I do think there's something about a DSL that would drive adoption for using deprecations internally in a way that dropping down one level to `ActiveSupport::Deprecation.warn` doesn't.

Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: Agreed 👍

Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: I'd propose combining those two bullets into a single "deprecate this method" helper, and then wrapping that with my proposed `alias_attribute` change -- basically the goal is so that the "deprecate this method" helper never _also_ aliases, but the "alias this method" helper can _also_ optionally deprecate the old one.

Main goal there is to try to solve the "what should do the aliasing" problem intuitively and sidestep the problem of how to frame/describe/name the `aliased_attribute` argument since it would no longer exist.

Comment (pat-whitrock) [reply lib/uncruft/deprecate_attribute.rb:6]: Love it!

So, we would have:

```ruby
deprecate_method :name, message: '`name` is deprecated. Users are now anonymous.'
```

and

```ruby
alias_attribute :nickname, :name, deprecation: "`name` is deprecated. Please use `nickname` or `formal_name` instead."
```

That sounds great to me. I could still maybe see having a `deprecate_attribute` method without the alias, just so we have a shorthand for:

```ruby
deprecate_method :name, message: '`name` is deprecated. Users are now anonymous.'
deprecate_method :name=, message: '`name` is deprecated. Users are now anonymous.'
```

by writing:

```ruby
deprecate_attribute :name, message: '`name` is deprecated. Users are now anonymous.'
```

Side note: If we do eventually decide we're comfortable with default deprecation messages, it would be pretty cool that this would work with the same ergonomics as `Memoist`:

```ruby
deprecate_method def name
  [first_name, middle_initial, last_name].compact.join(' ')
end
```

Comment (smudge) [reply lib/uncruft/deprecate_attribute.rb:6]: Worth noting -- `alias_attribute` also aliases the `?` methods, so `deprecate_attribute` could do the same!!!

**Comment by smudge** [comment]
I love the API proposed in this comment: https://github.com/Betterment/uncruft/pull/4#discussion_r461163648

I'm hoping it would be possible to do as part of this PR, but the wrapping of `alias_attribute` might get a little tricky since that is coming from somewhere within Rails. We could also consider wrapping `alias_method` too, but again it's a similar trick to patch that to accept an additional argument.

So if that isn't possible, I'd still propose changing the ergonomics a little and defining `deprecate_attribute` and `deprecate_method` such that they **don't alias anything** and just do what they say (deprecate a method and supply a message).

@pat-whitrock @yieldjessyield how does that sound?

File: spec/uncruft/deprecate_attribute_spec.rb (hunk 4 above)
Comment (yieldjessyield) [reply spec/uncruft/deprecate_attribute_spec.rb:23]: I can just break them into seperate specs! Good call!

File: README.md [hunk 6]
```diff
@@ -39,6 +39,24 @@ You can also incrementally add new warnings to the ignorefile as you encounter t
 RECORD_DEPRECATIONS=1 rspec path/to/my/failing/spec.rb
 ```
 
+## Deprecating Attributes
+
```
Comment (yieldjessyield) [inline README.md:43]: Here's my first pass at a README entry! I like how @smudge gave a real-world example on why one might want to record deprecations above ex: `When testing on a new Rails version for the first time, you will undoubtedly encounter many new warnings.` - curious if @pat-whitrock or @smudge  have some suggestions on adding a real-world example to this? Would that be helpful to add?

File: README.md [hunk 7]
```diff
@@ -39,6 +39,24 @@ You can also incrementally add new warnings to the ignorefile as you encounter t
...
+If you would like to deprecate an attribute by aliasing it to a new attribute and applying an `ActiveSupport::Deprecation` warning on the deprecated attribute's getters and setters then look no further, we have a tool for that! Simply include `Uncruft::DeprecateAttribute` in your class, identify the attribute you would like deprecated, the attribute you would like aliased, and finally a message you would like applied to the deprecation warning.
```
Comment (yieldjessyield) [inline README.md:44]: Please feel free to suggest edits to this description!

**Comment by yieldjessyield** [comment]
bump! - sorry I was out last week and finally getting around to this. Updated specs (thanks for the suggestions @pat-whitrock ) and updated the README. Would LOVE any edits or suggestions for the README 😅

**Comment by yieldjessyield** [comment]
bump

Ok! Removed the alias functionality, which could be added in a later PR. Right now just adding deprecation functionality and renaming it to `deprecate_method` Curious if that name makes sense vs calling it `deprecate_attribute`?

File: lib/uncruft/deprecate_method.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+module Uncruft
+  module DeprecateMethod
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def deprecate_method(method, message:)
```
Comment (smudge) [inline lib/uncruft/deprecate_method.rb:6]: I think this would still make sense with the prior name -- `deprecate_attribute` -- since it deprecates both the setter and the getter. But if you wanted, you could define it as this:

```ruby
deprecate_method method, message
deprecate_method :"#{method}=, message
```

So that's making use of a more generic `deprecate_method` method that could be defined something like this:

```ruby
prepended_method = Module.new
  define_method method do |*args, &block|
    ActiveSupport::Deprecation.warn(message)
    super(*args, &block) # passing all args straight through so that deprecation does not care about method arity
  end
  prepend prepended_method
end
```

If both `deprecate_method` and `deprecate_attribute` are public, then they can both be used by downstream consumers.

**Comment by smudge** [comment]
Love the README update!

I made one more suggestion about renaming back to `deprecate_attribute`, but potentially defining a second `deprecate_method` method. That could also be done in a separate PR, but if so I think we should go back to the `deprecate_attribute` naming still for the method that handles both the setter and the getter!

Aside from that, I think this is ready to ship -- LMK what you think!

platform TAFN
